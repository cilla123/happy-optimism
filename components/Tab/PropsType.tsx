export default interface PropsType {
  theme?: 'primary' | 'success' | 'warning' | 'error';
  lineWidth?: string | number;
  disabled?: boolean;
  canSwipe?: boolean;
  onChange?: (index: number) => void;
  children?: any;
}
