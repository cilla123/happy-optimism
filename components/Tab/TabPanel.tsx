import React, { PureComponent } from 'react';
import classnames from 'classnames';

export interface TabPannelProps {
  prefixCls?: string;
  className?: string;
}

export default class TabPanel extends PureComponent<TabPannelProps, any> {

  static defaultProps = {
    prefixCls: 'za-tab',
  };

  constructor(props) {
    super(props);
    this.state = {
      selected: props.selected || props.defaultSelected,
    };
  }

  componentWillReceiveProps(nextProps) {
    if ('selected' in nextProps) {
      this.setState({
        selected: !!nextProps.selected,
      });
    }
  }

  render() {
    const { prefixCls, className, children } = this.props;

    const cls = classnames(`${prefixCls}-panel-item`, className, {
      active: !!this.state.selected,
    });

    return <div className={cls} role="tabpanel">{children}</div>;
  }
}
