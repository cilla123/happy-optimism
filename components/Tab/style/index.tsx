import '../../style/';
import '../../Swipe/style/';
import './index.scss';
