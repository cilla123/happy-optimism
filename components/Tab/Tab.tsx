import React, { PureComponent, CSSProperties } from 'react';
import PropsType from './PropsType';
import classnames from 'classnames';
import TabPanel from './TabPanel';
import Swipe from '../Swipe';

const getSelectIndex = (children) => {
  let selectIndex;
  React.Children.forEach(children, (item: any, index) => {
    if (item.props && item.props.selected) {
      selectIndex = index;
    }
  });
  return selectIndex;
};

export interface TabProps extends PropsType {
  prefixCls?: string;
  className?: string;
}

export default class Tab extends PureComponent<TabProps, any> {

  static Panel: any;
  static defaultProps = {
    prefixCls: 'za-tab',
    theme: 'primary',
    disabled: false,
    canSwipe: false,
  };

  private swipe;

  constructor(props) {
    super(props);
    this.state = {
      value: props.value || props.defaultValue || getSelectIndex(props.children) || 0,
    };
  }

  componentWillReceiveProps(nextProps) {
    if ('value' in nextProps || getSelectIndex(nextProps.children)) {
      this.setState({
        value: nextProps.value || nextProps.defaultValue || getSelectIndex(nextProps.children) || 0,
      });
      if (typeof nextProps.onChange === 'function') {
        nextProps.onChange(nextProps.value);
      }
    }
  }

  onSwipeChange = (value) => {
    const { onChange } = this.props;
    this.setState({ value });
    if (typeof onChange === 'function') {
      onChange(value);
    }
  }

  onTabClick = (tab, index) => {
    const { disabled, canSwipe, onChange } = this.props;
    if (disabled || tab.props.disabled) {
      return;
    }
    this.setState({ value: index });
    if (typeof onChange === 'function') {
      onChange(index);
    }
    if (canSwipe) {
      this.swipe.onSlideTo(index);
    }
  }

  renderTabs = (tab, index) => {
    const { prefixCls, disabled } = this.props;
    const itemCls = classnames(`${prefixCls}-header-item`, tab.props.className, {
      disabled: disabled || tab.props.disabled,
      active: this.state.value === index,
      // hasline,
    });

    return (
      <li
        role="tab"
        key={+index}
        className={itemCls}
        onClick={() => this.onTabClick(tab, index)}
      >
        {tab.props.title}
      </li>
    );
  }

  render() {
    const { prefixCls, className, theme, lineWidth, canSwipe, children } = this.props;

    const classes = classnames(`${prefixCls}`, className, {
      [`theme-${theme}`]: !!theme,
    });

    // 渲染选项
    const tabsRender = React.Children.map(children, this.renderTabs);

    // 渲染内容
    let contentRender;

    if (canSwipe) {
      contentRender = (
        <Swipe
          direction="left"
          showPagination={false}
          activeIndex={this.state.value}
          ref={(ele) => { this.swipe = ele; }}
          onChange={(value) => this.onSwipeChange(value)}
        >
          {React.Children.map(children, (item: any) => <div>{item.props.children}</div>)}
        </Swipe>
      );
    } else {
      contentRender = React.Children.map(children, (item: any, index) => {
        return <TabPanel {...item.props} selected={this.state.value === index} />;
      });
    }

    const lineStyle: CSSProperties = {
      width: `${100 / children.length}%`,
      left: `${(this.state.value / children.length) * 100}%`,
      // right: `${(children.length - this.state.value - 1) / children.length * 100}%`,
      // transition: `right 0.3s cubic-bezier(0.35, 0, 0.25, 1), left 0.3s cubic-bezier(0.35, 0, 0.25, 1) 0.09s`,
    };

    let lineInnerRender;
    if (lineWidth) {
      lineStyle.backgroundColor = 'transparent';
      lineInnerRender = <span className={`${prefixCls}-line-inner`} style={{ width: lineWidth }} />;
    }

    return (
      <div className={classes}>
        <div className={`${prefixCls}-header`}>
          <ul role="tablist">{tabsRender}</ul>
          <div className={`${prefixCls}-line`} style={lineStyle}>{lineInnerRender}</div>
        </div>
        <div className={`${prefixCls}-container`}>
          {contentRender}
        </div>
      </div>
    );
  }
}
