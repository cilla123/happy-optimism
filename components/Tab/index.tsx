import Tab from './Tab';
import TabPanel from './TabPanel';

Tab.Panel = TabPanel;

export default Tab;
