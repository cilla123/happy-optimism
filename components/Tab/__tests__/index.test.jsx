import React from 'react';
import { render, shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import Tab from '../index';

describe('Tab', () => {
  it('renders correctly', () => {
    const wrapper = render(
      <Tab>
        <Tab.Panel title="选项卡1">
          <div>试试点我左滑</div>
        </Tab.Panel>
        <Tab.Panel title="选项卡2">
          <div>试试点我右滑</div>
        </Tab.Panel>
      </Tab>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('canSwipe', () => {
    const wrapper = render(
      <Tab canSwipe>
        <Tab.Panel title="选项卡1">
          <div>试试点我左滑</div>
        </Tab.Panel>
        <Tab.Panel title="选项卡2">
          <div>试试点我右滑</div>
        </Tab.Panel>
      </Tab>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('lineWidth is auto', () => {
    const wrapper = shallow(
      <Tab lineWidth="auto">
        <Tab.Panel title="选项卡1">
          <div>试试点我左滑</div>
        </Tab.Panel>
        <Tab.Panel title="选项卡2">
          <div>试试点我右滑</div>
        </Tab.Panel>
      </Tab>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
    wrapper.setProps({ lineWidth: 30 });
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('receive new value', () => {
    const onChange = jest.fn();
    const wrapper = shallow(
      <Tab onChange={onChange}>
        <Tab.Panel title="选项卡1">
          <div>试试点我左滑</div>
        </Tab.Panel>
        <Tab.Panel title="选项卡2">
          <div>试试点我右滑</div>
        </Tab.Panel>
      </Tab>
    );
    wrapper.setProps({ value: 1 });
  });

  it('click tabs', () => {
    const onChange = jest.fn();
    const wrapper = mount(
      <Tab canSwipe onChange={onChange}>
        <Tab.Panel title="选项卡1" disabled>
          <div>试试点我左滑</div>
        </Tab.Panel>
        <Tab.Panel title="选项卡2">
          <div>试试点我右滑</div>
        </Tab.Panel>
      </Tab>
    );
    wrapper.find('.za-tab-header-item').first().simulate('click');
    wrapper.find('.za-tab-header-item').last().simulate('click');
    expect(onChange).toBeCalledWith(1);
  });
});
