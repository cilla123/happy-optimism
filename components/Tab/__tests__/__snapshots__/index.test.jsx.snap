// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tab canSwipe 1`] = `
<div
  class="za-tab theme-primary"
>
  <div
    class="za-tab-header"
  >
    <ul
      role="tablist"
    >
      <li
        class="za-tab-header-item active"
        role="tab"
      >
        选项卡1
      </li>
      <li
        class="za-tab-header-item"
        role="tab"
      >
        选项卡2
      </li>
    </ul>
    <div
      class="za-tab-line"
      style="width:50%;left:0%"
    />
  </div>
  <div
    class="za-tab-container"
  >
    <div
      class="za-swipe"
    >
      <div
        class="za-swipe-items"
        style="white-space:nowrap"
      >
        <div
          class="za-swipe-item"
        >
          <div>
            试试点我左滑
          </div>
        </div>
        <div
          class="za-swipe-item"
        >
          <div>
            试试点我右滑
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Tab lineWidth is auto 1`] = `
<div
  className="za-tab theme-primary"
>
  <div
    className="za-tab-header"
  >
    <ul
      role="tablist"
    >
      <li
        className="za-tab-header-item active"
        key="0/.0"
        onClick={[Function]}
        role="tab"
      >
        选项卡1
      </li>
      <li
        className="za-tab-header-item"
        key="1/.1"
        onClick={[Function]}
        role="tab"
      >
        选项卡2
      </li>
    </ul>
    <div
      className="za-tab-line"
      style={
        Object {
          "backgroundColor": "transparent",
          "left": "0%",
          "width": "50%",
        }
      }
    >
      <span
        className="za-tab-line-inner"
        style={
          Object {
            "width": "auto",
          }
        }
      />
    </div>
  </div>
  <div
    className="za-tab-container"
  >
    <TabPanel
      key=".0"
      prefixCls="za-tab"
      selected={true}
      title="选项卡1"
    >
      <div>
        试试点我左滑
      </div>
    </TabPanel>
    <TabPanel
      key=".1"
      prefixCls="za-tab"
      selected={false}
      title="选项卡2"
    >
      <div>
        试试点我右滑
      </div>
    </TabPanel>
  </div>
</div>
`;

exports[`Tab lineWidth is auto 2`] = `
<div
  className="za-tab theme-primary"
>
  <div
    className="za-tab-header"
  >
    <ul
      role="tablist"
    >
      <li
        className="za-tab-header-item active"
        key="0/.0"
        onClick={[Function]}
        role="tab"
      >
        选项卡1
      </li>
      <li
        className="za-tab-header-item"
        key="1/.1"
        onClick={[Function]}
        role="tab"
      >
        选项卡2
      </li>
    </ul>
    <div
      className="za-tab-line"
      style={
        Object {
          "backgroundColor": "transparent",
          "left": "0%",
          "width": "50%",
        }
      }
    >
      <span
        className="za-tab-line-inner"
        style={
          Object {
            "width": 30,
          }
        }
      />
    </div>
  </div>
  <div
    className="za-tab-container"
  >
    <TabPanel
      key=".0"
      prefixCls="za-tab"
      selected={true}
      title="选项卡1"
    >
      <div>
        试试点我左滑
      </div>
    </TabPanel>
    <TabPanel
      key=".1"
      prefixCls="za-tab"
      selected={false}
      title="选项卡2"
    >
      <div>
        试试点我右滑
      </div>
    </TabPanel>
  </div>
</div>
`;

exports[`Tab renders correctly 1`] = `
<div
  class="za-tab theme-primary"
>
  <div
    class="za-tab-header"
  >
    <ul
      role="tablist"
    >
      <li
        class="za-tab-header-item active"
        role="tab"
      >
        选项卡1
      </li>
      <li
        class="za-tab-header-item"
        role="tab"
      >
        选项卡2
      </li>
    </ul>
    <div
      class="za-tab-line"
      style="width:50%;left:0%"
    />
  </div>
  <div
    class="za-tab-container"
  >
    <div
      class="za-tab-panel-item active"
      role="tabpanel"
    >
      <div>
        试试点我左滑
      </div>
    </div>
    <div
      class="za-tab-panel-item"
      role="tabpanel"
    >
      <div>
        试试点我右滑
      </div>
    </div>
  </div>
</div>
`;
